pilot/pkg/networking/core/v1alpha3/listener_builder.go

```diff
@@ -729,29 +727,23 @@ func buildOutboundCatchAllNetworkFiltersOnly(push *model.PushContext, node *mode
 	return filterStack
 }
 
-// TODO: This code is still insufficient. Ideally we should be parsing all the virtual services
-// with TLS blocks and build the appropriate filter chain matches and routes here. And then finally
-// evaluate the left over unmatched TLS traffic using allow_any or registry_only.
-// See https://github.com/istio/istio/issues/21170
-func buildOutboundCatchAllNetworkFilterChains(_ *ConfigGeneratorImpl,
-	node *model.Proxy, push *model.PushContext) []*listener.FilterChain {
-	filterStack := buildOutboundCatchAllNetworkFiltersOnly(push, node)
-	chains := make([]*listener.FilterChain, 0, 2)
-	chains = append(chains, blackholeFilterChain(push, node), &listener.FilterChain{
-		Name:    model.VirtualOutboundCatchAllTCPFilterChainName,
-		Filters: filterStack,
-	})
-	return chains
-}
-
-func blackholeFilterChain(push *model.PushContext, node *model.Proxy) *listener.FilterChain {
+func outboundBlackholeFilterChain(localhost string, push *model.PushContext, node *model.Proxy) *listener.FilterChain {
 	return &listener.FilterChain{
 		Name: model.VirtualOutboundBlackholeFilterChainName,
 		FilterChainMatch: &listener.FilterChainMatch{
-			// We should not allow requests to the listen port directly. Requests must be
-			// sent to some other original port and iptables redirected to 15001. This
-			// ensures we do not passthrough back to the listen port.
+			// Prevent traffic loops on the listener port while still allowing outbound passthrough
+			// to external endpoints that use this same port.
+			//
+			// This only block traffic loops of the form localhost:15001 -> localhost:15001.
+			// Traffic loops of the forms pod_ip:15001 -> pod_ip:15001, pod_ip:15006 -> pod_ip:15006,
+			// and localhost:15006 -> localhost:15006 are blocked on the inbound listener.
 			DestinationPort: &wrappers.UInt32Value{Value: uint32(push.Mesh.ProxyListenPort)},
+			PrefixRanges: []*core.CidrRange{
+				{
+					AddressPrefix: localhost,
```

Comment:
With the outbound blackhole filter chain scoped to only match localhost as destination, it allows outbound to non-localhost on 15001, but blocks localhost loops. This does not block loops on Pod IP, but those are handled by the inbound listener, which blocks _all_ traffic with original_dst on ports 15001 or 15006.